validator/rules/executor.go

```diff
@@ -25,21 +25,67 @@ func (config *RuleConfig) ExecuteRules(cmd *cobra.Command) []validator.Validatio
 	// initialize default rules
 	config.initDefaultRules()
 
+	// validate the root command
+	config.validate(cmd, &info)
+
 	return config.executeHelper(cmd, info)
 }
 
+// executeRulesChildren execute rules on children of cmd
+func (config *RuleConfig) executeRulesChildren(cmd *cobra.Command, info validator.StatusLog) []validator.ValidationError {
+
+	children := cmd.Commands()
+	for _, child := range children {
+
+		if !child.IsAvailableCommand() || child.IsAdditionalHelpTopicCommand() {
+			continue
+		}
+
+		config.validate(child, &info)
+	}
+	return info.Errors
+}
+
+// executeHelper recursively traverse over all the subcommands
+// and validate using executeRulesChildren function
 func (config *RuleConfig) executeHelper(cmd *cobra.Command, info validator.StatusLog) []validator.ValidationError {
 
-	return validator.Traverse(
-		cmd,
-		config.Verbose,
-		info,
-		func(cmd *cobra.Command, verbose bool) []validator.ValidationError {
-			info.Errors = append(info.Errors, validateLength(cmd, config.Length, config.Verbose)...)
-			info.Errors = append(info.Errors, validateMustExist(cmd, config.MustExist, config.Verbose)...)
-			return info.Errors
-		},
-	)
+	for _, child := range cmd.Commands() {
+
+		// base case
+		if !child.IsAvailableCommand() || child.IsAdditionalHelpTopicCommand() {
+			continue
+		}
+
+		// recursive call
+		info.Errors = config.executeHelper(child, info)
+
+	}
+
+	// prints additional info in debug mode
+	if config.Verbose {
+		log.Printf("commands checked: %d\nchecks failed: %d\n", info.TotalTested, info.TotalErrors)
+	}
+
+	info.Errors = config.executeRulesChildren(cmd, info)
+
+	return info.Errors
+}
+
+// validate returns validation errors by executing the rules
+func (config *RuleConfig) validate(cmd *cobra.Command, info *validator.StatusLog) {
+
+	// Length rule
+	lenErrs := validateLength(cmd, config.Length, config.Verbose)
```

Comment:
Lets use for loop for processing rules rather than calling each rule

Reply:
@wtrocki I have addressed above 2 comments. As all rules gives different function to execute, how can I use them in loop?
like length - validateLength
MustExist - validateMustExist

And rule config also looks something like this
```go
type RuleConfig struct {
	Verbose bool
	Length
	MustExist
}
```
Can you suggest some way?